boa/contracts/vyper/compiler_utils.py

```diff
@@ -104,43 +109,52 @@ def generate_bytecode_for_internal_fn(fn):
         _fn_sig.append(sig_arg_text)
     fn_sig = ", ".join(_fn_sig)
 
-    wrapper_code = textwrap.dedent(
+    return textwrap.dedent(
         f"""
         @external
         @payable
         def __boa_private_{fn_name}__({fn_sig}){return_sig}:
             {fn_call}
     """
     )
-    return compile_vyper_function(wrapper_code, contract)
 
 
 def generate_bytecode_for_arbitrary_stmt(source_code, contract):
     """Wraps arbitrary stmts with external fn and generates bytecode"""
+    wrapper_code = generate_source_for_arbitrary_stmt(source_code, contract)
+    return compile_vyper_function(wrapper_code, contract)
 
-    ast = parse_to_ast(source_code)
-    vy_ast.folding.fold(ast)
-    ast = ast.body[0]
 
-    return_sig = ""
-    debug_body = source_code
+def generate_source_for_arbitrary_stmt(source_code, contract):
+    """Wraps arbitrary stmts with external fn and generates source code"""
 
-    if isinstance(ast, vy_ast.Expr):
-        with contract.override_vyper_namespace():
-            try:
-                ast_typ = get_exact_type_from_node(ast.value)
-                return_sig = f"-> {ast_typ}"
-                debug_body = f"return {source_code}"
-            except InvalidType:
-                pass
+    ast_typ = detect_expr_type(source_code, contract)
+    if ast_typ:
+        return_sig = f"-> {ast_typ}"
+        debug_body = f"return {source_code}"
+    else:
+        return_sig = ""
+        debug_body = source_code
 
     # wrap code in function so that we can easily generate code for it
-    wrapper_code = textwrap.dedent(
+    return textwrap.dedent(
         f"""
         @external
         @payable
         def __boa_debug__() {return_sig}:
             {debug_body}
     """
     )
-    return compile_vyper_function(wrapper_code, contract)
+
+
+def detect_expr_type(source_code, contract):
```

Comment:
i'm not really sure why this needs to be refactored out

Reply:
We use this function directly for zksync eval

Reply:
🤣 not sure.. thats for another review tho
This refactor is just separating concerns

Reply:
i don't think we should separate concerns until the need for it is demonstrated. if we aren't decided on whether the zksync plugin should have it, maybe the best solution for now is to implement a separate copy of `detect_expr_type` within the zksync plugin

Reply:
I still like the refactor but the code isn't too long, i can copy it